### DISCUSSION TITLE
Entanglement entropy in 1d extended bose hubbard model

### DISCUSSION THREAD

**Discussion by KoutaDagnino** [QuSpin/QuSpin#447] [Q&A] [answered]
Dear QuSpin team,

I had a couple questions about how one could go around calculating entanglement entropies in QuSpin. Looking at the documentation I saw that the standard way is using basis.ent_entropy. Thus I tried computing the entanglement entropy in a 1D extended bose hubbard model:

`from quspin.operators import hamiltonian # Hamiltonians and operators
from quspin.basis import boson_basis_1d # bosonic Hilbert space
import matplotlib.pyplot as plt

L = 14 # length of chain
N = 7 # number of bosons
t = 1 # hopping coefficient
V=2 # n.n. site coupling 
U = -4 # Hubbard interaction
#

int_list_1 = [[-0.5*U,i] for i in range(L)] # interaction $-U/2 \sum_i n_i$
int_list_2 = [[0.5*U,i,i] for i in range(L)] # interaction: $U/2 \num_i n_i^2$

hop_list = [[-t,i,(i+1)%L] for i in range(L)] # hopping terms using PBCs

ext_list = [[V,i,(i+1)%L] for i in range(L)] # adjacent site coupling terms using PBCs

static = [
            ["+-",hop_list], # hopping
            ["-+",hop_list],
            ["nn",int_list_2], # U n_i^2
            ["n",int_list_1], # -U n_i
            ["nn",ext_list], # V n_i n_j
        ]

dynamic = [] # no dynamic operators
    
basis = boson_basis_1d(L, Nb=N,kblock=0)

h = hamiltonian(static, dynamic, basis=basis, check_symm=False)

E,states=h.eigh()

op_static=[["nn", [[1/L,i,(i+1)%L] for i in range(L)]]]

op = hamiltonian(op_static,[],basis=basis, check_symm=False)

deg=states.shape[1]

Entropy=[]

for i in range(deg):
    subsystem=[j for j in range(L//2)] # sites contained in subsystem
    Sent=basis.ent_entropy(states[:,i],sub_sys_A=subsystem)["Sent_A"]
    Entropy.extend([Sent]) 

plt.scatter(E,Entropy,marker="o",s=3,color="k")
plt.grid(linewidth=0.1)
plt.xlabel("Energy")
plt.ylabel("Entanglement entropy")
plt.show()
`
This seemed to work with N=4, L=8 but with larger systems it seems to not compile (kernel gets killed). 

I tried doing the same for a quantum scarred PXP model with N=18 (hilbert space dimension of 5700) and quspin was able to calculate the entropy for the entire spectrum despite having the same Hilbert space as the previous model. Here is the code (taken from http://weinbe58.github.io/QuSpin/examples/example14.html):

`from quspin.operators import hamiltonian
from quspin.basis import spin_basis_1d # Hilbert space spin basis_1d
from quspin.basis.user import user_basis # Hilbert space user basis
from quspin.basis.user import pre_check_state_sig_32,op_sig_32,map_sig_32 # user basis data types
from numba import carray,cfunc # numba helper functions
from numba import uint32,int32 # numba data types
import numpy as np
import matplotlib.pyplot as plt
#
N = 18 # total number of lattice sites
#
######  function to call when applying operators
@cfunc(op_sig_32, locals=dict(s=int32,b=uint32))
def op(op_struct_ptr,op_str,ind,N,args):
    # using struct pointer to pass op_struct_ptr back to C++ see numba Records
    op_struct = carray(op_struct_ptr,1)[0]
    err = 0
    ind = N - ind - 1 # convention for QuSpin for mapping from bits to sites.
    s = (((op_struct.state>>ind)&1)<<1)-1
    b = (1<<ind)
    #
    if op_str==120: # "x" is integer value 120 (check with ord("x"))
        op_struct.state ^= b
    elif op_str==121: # "y" is integer value 120 (check with ord("y"))
        op_struct.state ^= b
        op_struct.matrix_ele *= 1.0j*s
    elif op_str==122: # "z" is integer value 120 (check with ord("z"))
        op_struct.matrix_ele *= s
    else:
        op_struct.matrix_ele = 0
        err = -1
    #
    return err
#
op_args=np.array([],dtype=np.uint32)
#
######  function to filter states/project states out of the basis
#
@cfunc(pre_check_state_sig_32,
    locals=dict(s_shift_left=uint32,s_shift_right=uint32), )
def pre_check_state(s,N,args):
    """ imposes that that a bit with 1 must be preceded and followed by 0,
    i.e. a particle on a given site must have empty neighboring sites.
    #
    Works only for lattices of up to N=32 sites (otherwise, change mask)
    #
    """
    mask = (0xffffffff >> (32 - N)) # works for lattices of up to 32 sites
    # cycle bits left by 1 periodically
    s_shift_left = (((s << 1) & mask) | ((s >> (N - 1)) & mask))
    #
    # cycle bits right by 1 periodically
    s_shift_right = (((s >> 1) & mask) | ((s << (N - 1)) & mask))
    #
    return (((s_shift_right|s_shift_left)&s))==0
#
pre_check_state_args=None
#
######  construct user_basis 
# define maps dict
maps = dict() # no symmetries to apply.
# define op_dict
op_dict = dict(op=op,op_args=op_args)
# define pre_check_state
pre_check_state=(pre_check_state,pre_check_state_args) # None gives a null pinter to args
# create user basis
basis = user_basis(np.uint32,N,op_dict,allowed_ops=set("xyz"),sps=2,
                    pre_check_state=pre_check_state,Ns_block_est=900000,**maps)
# print basis
print(basis)
#
###### construct Hamiltonian
# site-coupling lists
h_list  = [[1.0,i] for i in range(N)]
# operator string lists
static = [["x",h_list],]
# compute Hamiltonian, no checks have been implemented
no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
H = hamiltonian(static,[],basis=basis,dtype=np.float64,**no_checks)
E,states=H.eigh()

deg=states.shape[1]
Entropy=[]
for i in range(deg):
    subsystem=[i for i in range(N//2)]
    entr=basis.ent_entropy(states[:,i],sub_sys_A=subsystem)['Sent_A']
    Entropy.extend([entr])

plt.scatter(E,Entropy,marker="o",s=3,color="k")
plt.grid(linewidth=0.1)
plt.xlabel("Energy")
plt.ylabel("Entanglement entropy")
plt.show()
`

So is there something specific to the 1 chain model (or the PXP model) that makes it much harder (or easier) to compute the entanglement entropy?

EDIT:
Here are the codes in txt form:
[1D EBHM.txt](https://github.com/weinbe58/QuSpin/files/7093181/1D.EBHM.txt)
[PXP model.txt](https://github.com/weinbe58/QuSpin/files/7093183/PXP.model.txt)

**Comment by mgbukov** [comment]
@KoutaDagnino can you upload the two scripts as .txt files here? This way they're easier to run.

**Reply by KoutaDagnino** [reply]
I've edited my question with the txt files. I hope that helps.

**Comment by mgbukov** [comment, marked as answer]
One notable difference is that in the boson model you use symmetries. This makes the calculation of Sent more involved because quspin first lifts every state from the symmetry-reduced sector to the full H_space (w/o symmetries), and then performs the partial trace. There is no way around this that we know of, because the two subsystems are only well-defined in the original Fock basis, and not in the basis of the reduced H-space. When the original Fock basis is too large, this leads to limitations in computing Sent. For comparison, the original Fock space for bosons is $N^L$, while for spins its $2^L$, so you can check where the two are equal and quspin should behave the same speed-wise when computing Sent.  

Unfortunately, there's not much one can do here to help: you're fighting an exponential and this is not a fight that a classical computer can win.

**Reply by KoutaDagnino** [reply]
I've noticed however that for the boson model at N=5,L=10 the total hilbert space size is 2000, while the reduced symmetry space size is 200. When I try to compute the entanglement entropy for just 1 eigenstate in this model (within the k=0 sector) it takes far longer than when I compute the entanglement for 5000 eigenstates of the PXP model (with total hilbert space of 5000). This happens even if I already give quspin the state in the full hilbert space.

**Comment by mgbukov** [comment]
for the boson model at N=5,L=10 the total Hilbert space size is $N^L= 5**10=9 765 625$. You can reduce it to 2002 if you take particle conservation which is a symmetry.

 It might be possible to compute Sent within the fixed-particle sector, though. Quspin's `basis.ent_entropy()` cannot do this currently but if you're willing to spend time and write your own code, I would be happy to help. 

To explain where the problem comes from, consider a spin-1/2 system with magnetization (i.e. particle number) conservation: you have 4 lattice sites with 2  spin-up's and 2 spin-down's in every basis state. Because of the magnetization conservation, the total H-space dimension for the full system is 4 choose 2 = 6 states (as compared to 2**4 = 16 states in the full H-space). Now let's say you want to compute Sent w.r.t. the first two lattice sites: you will notice that even though magnetization is conserved on the total system of L sites, it is not conserved on the subsystem. Hence, when performing the partial trace you have to go from a basis with magnetization conservation on the full system to a basis that does not have magnetization conservation on the subsystem. It is possible to compute Sent in this situation, but it requires some extra effort and `basis.ent_entropy()` is meant as a general function so it does not handle this case. 

Let me know if you intend to write a function to do this yourself and I'll refer you to some useful quspin functions to get you started.

**Reply by KoutaDagnino** [reply]
Oh I see thank you for the very clear explanation! Could you go more into detail as to how one could calculate the entropy within a pcon sector?
EDIT: perhaps some sort of block diagonalisation of the density matrix?

**Comment by mgbukov** [comment]
Well, this is what you have to figure out (if it's possible in the first place), but it likely is. I don't know how to do it off the top of my head, but we did discuss this with @weinbe58 long time ago; maybe he remembers some details? 

The good news for you is that quspin has the function `basis.project_from(...,pcon=True)` which will take a state in the symmetry basis (including momentum etc.) and will lift it to a state in the particle-number basis. I don't remember but I think this is done without going thru the full basis first for the basis_general classes (need to check this).

**Reply by KoutaDagnino** [reply]
I see I'll definitely look into it thanks!

**Reply by mgbukov** [reply]
great! keep us posted how this goes.
